components/rating/rating.tsx

```diff
@@ -0,0 +1,124 @@
+  value?: 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9
+  valueCallback?: React.Dispatch<React.SetStateAction<number>>
+  lock?: boolean
+  lockCallback?: React.Dispatch<React.SetStateAction<boolean>>
+  onClick?: React.MouseEventHandler<SVGElement>
+  onMouseEnter?: React.MouseEventHandler<SVGElement>
+}
+
+const defaultProps = {
+  type: 'default' as NormalTypes,
+  className: '',
+  icon: (props: any): React.ReactElement<Icon> => <Star {...props} />,
+  count: 5,
+  value: 0,
+  valueCallback: () => {},
+  lock: false,
+  lockCallback: () => {},
+  onClick: () => {},
+  onMouseEnter: () => {},
+}
+
+type NativeAttrs = Omit<React.HTMLAttributes<any>, keyof Props>
+export type RatingProps = Props & typeof defaultProps & NativeAttrs
+
+const getColor = (type: NormalTypes, theme: GeistUIThemes): string => {
+  const colors: { [key in NormalTypes]?: string } = {
+    default: theme.palette.foreground,
+    success: theme.palette.success,
+    warning: theme.palette.warning,
+    error: theme.palette.error,
+  }
+  return colors[type] || (colors.default as string)
+}
+
+const Rating: React.FC = ({
+  type,
+  children,
+  className,
+  icon,
+  count,
+  value,
+  valueCallback,
+  lock,
+  lockCallback,
+  onClick,
+  onMouseEnter,
+  ...props
+}: RatingProps) => {
+  const theme = useTheme()
+  const color = useMemo(() => getColor(type, theme), [type, theme])
+  const [hoverState, setHoverState] = useState<number>(value) // state is from 0 to count - 1
+  const [isLocked, setIsLocked] = useState<boolean>(lock)
+
+  useEffect(() => {
+    valueCallback(hoverState + 1) // state + 1
+  }, [hoverState])
+
+  useEffect(() => {
+    lockCallback(isLocked)
+  }, [isLocked])
+
+  const handleMouseUp = (event: React.MouseEvent<SVGElement>, index: number) => {
+    if (isLocked) {
+      setIsLocked(false) // unlock
+      onClick && onClick(event)
+      return
+    }
+    setHoverState(index)
+    setIsLocked(true) // lock
+    onClick && onClick(event)
+  }
+
+  const handleMouseEnter = (event: React.MouseEvent<SVGElement>, index: number) => {
+    if (isLocked) return // leave on lock
+    setHoverState(index)
+    onMouseEnter && onMouseEnter(event)
+  }
+
+  const Icon = icon
+
+  const render = () => {
+    return (
+      <>
+        {[...Array(count)].map((_e, i) => (
+          <Icon
+            key={i}
+            color={color}
+            className={`${className}`}
+            fill={i <= hoverState ? color : 'transparent'}
+            stroke={i <= hoverState ? '#fff' : color}
+            transform={i <= hoverState ? 'scale(1.1)' : 'scale(1)'}
+            onMouseEnter={(e: React.MouseEvent<SVGElement, MouseEvent>) =>
+              handleMouseEnter(e, i)
+            }
+            onMouseUp={(e: React.MouseEvent<SVGElement, MouseEvent>) =>
+              handleMouseUp(e, i)
+            }
+            {...props}>
+            {children}
+          </Icon>
+        ))}
+      </>
+    )
+  }
+  return render()
```

Comment:
`render` should not be necessary, it is possible to return anonymous functions directly.

Reply:
Good Catch 👍